Add osu!stable style circular song progress for legacy skins

LeNitrous opened the pull request:
Resolves https://github.com/ppy/osu/issues/10945.

Some feedback concerns the current song progress which takes up too much space even with the difficulty graph contracted and they prefer stable's song progress.

This PR intends to replace the currently default song progress and make use of the pie chart styled progress for legacy skins. With its minimal appearance, controls for seeking are not available for this style. Although it is still possible through key binds.

Everything is eyeballed to what I think is close enough. Functionality should be on parity with stable.

https://user-images.githubusercontent.com/20495991/181137622-8032e84c-c83b-48d0-95f3-4c055f4c4462.mp4

![image](https://user-images.githubusercontent.com/20495991/181137597-4e7028e1-f25c-4c83-8542-08df75ae8bad.png)

LeNitrous committed:
implement `LegacySongProgress`

LeNitrous committed:
replace `SongProgress` with `LegacySongProgress`

LeNitrous committed:
Merge branch 'master' into pie-chart-progress

LeNitrous committed:
remove unused using

peppy committed:
Remove unused `combo` DI retrieval

peppy committed:
Transfer TODO comment across to copy-pasted implmentation

peppy commented:
How are you testing this? I can't find a test scene which shows this control in action..

peppy committed:
Fix broken tests due to badly reimplemented copy-pasted code

peppy reviewed (changes requested):
No tests

peppy commented:
A quick visual pass in gameplay shows a few inconsistencies:

- To make colours match stable:

```csharp
            progress.Colour = progress.Progress < 0 ?
                new Color(199, 255, 47, 153) :
                new Color(255, 255, 255, 153);
```

- The fill is too large and can be seen outside the border. It should be reduced by 0.5 or 1 pixel to avoid that.

LeNitrous commented:
Looking into simplifying it since it has shared logic with `SongProgress`.

LeNitrous committed:
scale down graph from bleeding through border

LeNitrous committed:
make `SongProgress` abstract
- move unrelated logic to `DefaultSongProgress`
- make `LegacySongProgress` inherit `SongProgress`

LeNitrous committed:
move song progress graph to its own test scene

LeNitrous committed:
refactor song progress test scene

LeNitrous committed:
remove unused using

frenzibyte commented:
The progress bar doesn't cycle properly on intro:

https://user-images.githubusercontent.com/22781491/181225308-a7715b4f-feb6-43ea-90a9-d5e7114e5cda.mp4

Here's how stable looks on the same [beatmap](https://osu.ppy.sh/beatmapsets/393663#osu/865491), for reference:

https://user-images.githubusercontent.com/22781491/181229846-a9438a27-7af4-4057-950c-d6176f3e7e40.mp4

LeNitrous commented:
For starters, stable seems to have an earlier start time for beatmaps without storyboard intros (lazer's is -2000ms). Additionally, it may be because the graph relies on frame stable time. I'm getting smoother results on my end as opposed to your video.

frenzibyte commented:
I'm not sure how that's related to the issue here? As per the video above, the pie stays frozen until the clock reaches start time of first hitobject. And by looking at the code, it's probably due to relying on storyboard time solely, rather than the final start time value which is stored in `GameplayClockContainer.StartTime`.

LeNitrous committed:
account for gameplay start time

LeNitrous commented:
Seems like I wasn't accounting for the gameplay start time. Should be corrected now.

frenzibyte commented on osu.Game/Screens/Play/HUD/SongProgress.cs:
```diff
@@ -0,0 +1,90 @@
+
+            if (frameStableTime < FirstHitTime)
+            {
+                double earliest = Math.Min(FirstEventTime, gameplayClockContainer?.StartTime ?? 0);
+                UpdateProgress((frameStableTime - earliest) / (FirstHitTime - earliest), gameplayTime, true);
```
Why is `FirstEventTime` still required? `GameplayClockContianer.StartTime` should always replace it.

peppy replied:
I can't read this math. I don't know what it's doing. Split it out and comment.

peppy committed:
Split out tests and fix variable conflict

peppy commented on osu.Game.Tests/Visual/Gameplay/TestSceneSongProgress.cs:
```diff
@@ -135,10 +70,12 @@ private void displayMaxValues()
         {
-            progress.Objects = objects;
-            graph.Objects = objects;
+            defaultProgress.RequestSeek = pos => clock.Seek(pos);
```
Why is this in a method called `replaceObjects`?

peppy commented on osu.Game.Tests/Visual/Gameplay/TestSceneSongProgress.cs:
```diff
@@ -7,23 +7,21 @@
 {
     [TestFixture]
-    public class TestSceneSongProgress : OsuTestScene
+    public class TestSceneSongProgress : SkinnableHUDComponentTestScene
```
No test coverage of intro time?

peppy commented on osu.Game/Screens/Play/HUD/SongProgress.cs:
```diff
@@ -0,0 +1,90 @@
+            if (objects == null)
+                return;
+
+            double gameplayTime = gameplayClockContainer?.GameplayClock.CurrentTime ?? gameplayClock?.CurrentTime ?? Time.Current;
+            double frameStableTime = referenceClock?.CurrentTime ?? gameplayTime;
```
I have no idea why there are two times, and why they are using so many fallback paths.

Simplify this and add comments if they are both required.

peppy commented on osu.Game/Screens/Play/HUD/SongProgress.cs:
```diff
@@ -0,0 +1,90 @@
+    {
+        public bool UsesFixedAnchor { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private GameplayClock gameplayClock { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private GameplayClockContainer gameplayClockContainer { get; set; }
```
There's no way you need both of these.

frenzibyte commented on osu.Game/Screens/Play/HUD/SongProgress.cs:
```diff
@@ -0,0 +1,90 @@
+
+        protected double FirstEventTime { get; private set; }
+
+        protected abstract void UpdateProgress(double progress, double time, bool isIntro);
```
It's confusing to have `progress` and `time` together, since the former relies on frame stable time while the latter relies on gameplay clock time. `DefaultSongProgress` should just use `gameplayClock.CurrentTime` directly.

frenzibyte commented on osu.Game/Screens/Play/HUD/SongProgress.cs:
```diff
@@ -0,0 +1,90 @@
+        protected double FirstEventTime { get; private set; }
+
+        protected abstract void UpdateProgress(double progress, double time, bool isIntro);
+        protected abstract void UpdateObjects(IEnumerable<HitObject> objects);
```
I would rather combine this with the `Objects` property by making it virtual, rather than have a generic `UpdateObjects` method.

frenzibyte replied:
Please refrain from marking conversations as resolved if some of them aren't actually applied.

peppy replied:
Don't ever self-mark resolved.

frenzibyte commented on osu.Game.Tests/Visual/Gameplay/TestSceneSongProgress.cs:
```diff
@@ -147,15 +84,29 @@ protected override void Update()
+        protected override Drawable CreateDefaultImplementation()
         {
-            public int CreationCount { get; private set; }
+            defaultProgress = new DefaultSongProgress
+            {
+                RelativeSizeAxes = Axes.X,
+                Anchor = Anchor.BottomLeft,
+                Origin = Anchor.BottomLeft,
+            };
+
+            progresses.Add(defaultProgress);
+            return defaultProgress;
```
Use `this.ChildrenOfType<DefaultSongProgress>().Single()` and `this.ChildrenOfType<SongProgress>().ForEach` rather than storing references.

frenzibyte reviewed (changes requested):
In addition to the above.

frenzibyte commented on osu.Game/Screens/Play/HUD/SongProgress.cs:
```diff
@@ -0,0 +1,90 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
+
```
Also please enable NRT on all new classes.

LeNitrous committed:
apply suggestions
- refactor `SongProgress`
- make`UpdateProgress` more readable
- enable NRT on new classes
- refactor `TestSceneSongProgress` to use `GameplayClockContainer`

LeNitrous committed:
simplify helper methods

LeNitrous committed:
set `FirstHitTime` and `LastHitTime` once

LeNitrous committed:
apply code quality fixes

peppy committed:
Tidy up various things everywhere

peppy committed:
Fix `DefaultSongProgress` graph not resetting if time is in intro

peppy committed:
Remove pointless null check

peppy committed:
Move initial show to base implementation and add transition for legacy version

peppy committed:
Apply nullability to `DefaultSongProgress` and clean up more stuff

peppy committed:
Adjust metrics

peppy commented:
@frenzibyte Please re-review for sanity.

Kinda depends on https://github.com/ppy/osu/pull/19441 i guess.

peppy committed:
Merge branch 'master' into pie-chart-progress

peppy committed:
Remove unnecessary `LoadTrack` call

frenzibyte committed:
Remove redundant/overwritten specifications

frenzibyte committed:
Expose `StartTime` from gameplay clock

frenzibyte committed:
Remove caching of `GameplayClockContainer` in favour of `GameplayClock`

Also fixes `SongProgress` being displayed in skin editor on non-gameplay
screens, due to `GameplayClock` not marked as a required dependency.

frenzibyte commented:
Noticed the song progress components showing up in non-gameplay screens, and that was due to the dependency on `GameplayClockContainer` being optional instead of required. 

Making it required broke many test scenes and needed to create a container for each of them, so I went the other way around and exposed `StartTime` via `GameplayClock`, and also removed the cached attribute from `GameplayClockContainer` while at it.

frenzibyte committed:
Merge branch 'master' into pie-chart-progress

toadsworth-jr commented:
Would this PR retroactively remove the difficulty bar from previously imported scenes?
